src/main/java/edu/cornell/kfs/module/cg/document/CuAgencyMaintainableImpl.java

```diff
@@ -0,0 +1,26 @@
+package edu.cornell.kfs.module.cg.document;
+
+import org.kuali.kfs.module.cg.businessobject.Agency;
+import org.kuali.kfs.module.cg.document.AgencyMaintainableImpl;
+
+import edu.cornell.kfs.module.cg.businessobject.AgencyExtendedAttribute;
+
+public class CuAgencyMaintainableImpl extends AgencyMaintainableImpl {
+
+    private static final long serialVersionUID = 3088332201792805640L;
+
+    /**
+     * When CGB is enabled and the document is entering PROCESSED state, AgencyMaintainableImpl does
+     * some custom processing that will prematurely save the agency before performing any auto-setup
+     * of keys on the referenced objects. Thus, the agency extension needs to have its agency number
+     * manually set here, to avoid errors when persisting it.
+     */
+    @Override
+    public void prepareForSave() {
+        super.prepareForSave();
+        Agency agency = (Agency) getDataObject();
+        AgencyExtendedAttribute agencyExtension = (AgencyExtendedAttribute) agency.getExtension();
+        agencyExtension.setAgencyNumber(agency.getAgencyNumber());
```

Comment:
should there be error handling if the agency doesn't have an agency number yet?  Such as in the initial save scenario?

Reply:
The Agency Number is a required field, so it should have a value once the document is successfully submitted for routing. Prior to routing, the agency extension will only be stored in the maintenance XML and not in the agency extension table, so I don't think a null agency number would break anything here.

Reply:
OK, thanks for clarifying